staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
+			closeTransformers()
 			return err
 		}
-		s.transformerOverrides = transformerOverrides
-		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
+
+		var kmsPluginHealthzCheck healthz.HealthChecker
+		if len(encryptionConfiguration.HealthChecks) > 0 {
+			kmsPluginHealthzCheck = encryptionConfiguration.HealthChecks[0]
+		}
+
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers, kmsPluginHealthzCheck, closeTransformers, encryptionConfiguration.KMSTimeout)
+
+		// enable kms hot reload controller only if the config file is set to be automatically reloaded
+		if s.EncryptionProviderConfigAutomaticReload {
+			// with reload=true we will always have 1 health check
+			if len(encryptionConfiguration.HealthChecks) == 1 {
+				// add post start hook to start hot reload controller
+				// adding this hook here will ensure that it gets configured exactly once
+				err = addPostStartHook(
+					"start-encryption-provider-config-automatic-reload",
+					func(hookContext server.PostStartHookContext) error {
+						// get context
+						// we ignore the cancel func because this context should only be canceled when stopCh is closed
+						ctx, _ := wait.ContextForChannel(hookContext.StopCh)
+
+						kmsConfigController := kmsconfigcontroller.NewDynamicKMSEncryptionConfiguration(
+							"kms-encryption-config",
+							s.EncryptionProviderConfigFilepath,
+							s.dynamicTransformers,
+							encryptionConfiguration.EncryptionFileContentHash,
+							stopCh,
+						)
+
+						go kmsConfigController.Run(ctx)
+
+						return nil
+					},
+				)
+				if err != nil {
+					klog.Errorf("failed to add post start hook for kms encryption config hot reload controller: %v", err)
+					return err
+				}
+			} else {
+				klog.Errorf("failed to start kms encryption config hot reload controller. only 1 health check should be available when reload is enabled")
```

shouldn't this return an error rather than log?